base_layer/wallet_ffi/src/lib.rs

```diff
@@ -491,6 +491,51 @@ pub unsafe extern "C" fn public_key_to_emoji_node_id(pk: *mut TariPublicKey, err
     result = CString::new(emoji.as_str()).unwrap();
     CString::into_raw(result)
 }
+
+/// Creates a TariPublicKey from a char array in emoji format
+///
+/// ## Arguments
+/// `const *c_char` - The pointer to a TariPublicKey
+/// `error_out` - Pointer to an int which will be modified to an error code should one occur, may not be null. Functions
+/// as an out parameter.
+///
+/// ## Returns
+/// `*mut c_char` - Returns a pointer to a TariPublicKey. Note that it returns null on error.
+///
+/// # Safety
+#[no_mangle]
+pub unsafe extern "C" fn emoji_id_to_public_key(emoji: *const c_char, error_out: *mut c_int) -> *mut TariPublicKey {
+    let mut error = 0;
+    ptr::swap(error_out, &mut error as *mut c_int);
+    if emoji.is_null() {
+        error = LibWalletError::from(InterfaceError::NullError("emoji".to_string())).code;
+        ptr::swap(error_out, &mut error as *mut c_int);
+        return ptr::null_mut();
+    }
+
+    let emoji_str = CStr::from_ptr(emoji).to_str();
```

Comment:
```suggestion
    CStr::from_ptr(emoji).to_str()
    .and_then(|s| {
        EmojiId::str_to_pubkey(s)    
    })
    .and_then(|pk| {
         Box::into_raw(Box::new(pk))   
    }).ok_or_else(|_| {
                    error = LibWalletError::from(InterfaceError::InvalidEmojiId).code;
                    ptr::swap(error_out, &mut error as *mut c_int);
                    ptr::null_mut()    
    });
```
... and remove L#517-536. Might not be 100%, but the idea is there.
This only returns one error type (invalidID), but that's fine; and is quite a bit cleaner.